File: chisel_js/src/chisel/checker/services/svelte/concurrency.ts
import type { ProjectInfo } from "chisel/checker/models/project_info";
import type { Violation } from "chisel/checker/models/violation";
import { Severity } from "chisel/checker/models/severity";
import { createViolation } from "chisel/checker/models/violation";

export class ConcurrencyService {
  readonly ruleIdPrefix = "concurrency";

  check(project: ProjectInfo): Violation[] {
    const violations: Violation[] = [];
    for (const file of project.files) {
      if (!file.source) continue;
      const name = file.path;
      if (name.endsWith("+page.server.ts") || name.endsWith("+layout.server.ts")) {
        violations.push(...this._checkPromiseAll(file));
      }
    }
    return violations;
  }

  private _checkPromiseAll(file: { path: string; source: string }) {
    const violations: Violation[] = [];
    const lines = file.source.split("\n");
    for (let i = 0; i < lines.length; i++) {
      if (/Promise\.all\s*\(/.test(lines[i])) {
        violations.push(createViolation({
          file: file.path, line: i + 1, severity: Severity.WARNING,
          ruleId: "concurrency:promise-all-warning",
          message: "Promise.all across multiple services should only be used inside controllers, not directly in loaders.",
        }));
      }
    }
    return violations;
  }

  describeRules() {
    return [
      { id: "concurrency:promise-all-warning", category: "concurrency",
        description: "Promise.all used directly in a loader",
        fixGuidance: "Move Promise.all into a controller method. Single-service controllers are an anti-pattern — call service directly from loader via factory." },
    ];
  }
}

File: src/chisel/checker/services/concurrency.py
from __future__ import annotations

import ast
from dataclasses import dataclass

from chisel.checker.models.file_info import FileInfo
from chisel.checker.models.project_info import ProjectInfo
from chisel.checker.models.severity import Severity
from chisel.checker.models.violation import Violation


@dataclass(slots=True)
class ConcurrencyService:
    rule_id_prefix: str = "concurrency"

    def check(self, project: ProjectInfo) -> list[Violation]:
        violations: list[Violation] = []
        for file in project.files:
            if file.ast_tree is None:
                continue
            violations.extend(self._check_file(file))
        return violations

    def _check_file(self, file: FileInfo) -> list[Violation]:
        tree = file.ast_tree
        if tree is None:
            return []

        for node in ast.walk(tree):
            if isinstance(node, ast.Call) and isinstance(node.func, ast.Attribute):
                if isinstance(node.func.value, ast.Name):
                    if node.func.value.id == "asyncio" and node.func.attr == "gather":
                        return self._v(
                            file, node.lineno, "asyncio-gather-banned",
                            "asyncio.gather is banned — use asyncio.TaskGroup only",
                        )
        return []

    def _v(
        self, file: FileInfo, line: int, rule_suffix: str, message: str
    ) -> list[Violation]:
        return [
            Violation(
                file=str(file.path),
                line=line,
                severity=Severity.ERROR,
                rule_id=f"{self.rule_id_prefix}:{rule_suffix}",
                message=message,
            )
        ]
